### PR TITLE
refactor: email placeholder, additional content support and formatting

### PR DESCRIPTION
**Pull request by shohag121**
### All Submissions:

* [ ] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [ ] My code satisfies feature requirements
* [ ] My code is tested
* [ ] My code passes the PHPCS tests
* [ ] My code has proper inline documentation
* [ ] I've included developer documentation (optional)
* [ ] I've added proper labels to this pull request

### Changes proposed in this Pull Request:

* Closes https://github.com/weDevsOfficial/dokan-pro/issues/2465 (Pertially) 
* also Closes https://github.com/getdokan/dokan/issues/2010

### How to test the changes in this Pull Request:

1. check issue https://github.com/weDevsOfficial/dokan-pro/issues/2465

### Changelog entry

No change log needed

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

### CODE REVIEW

File: includes/Emails/ReverseWithdrawalInvoice.php [hunk 1]
```diff
@@ -43,23 +43,25 @@ class ReverseWithdrawalInvoice extends WC_Email {
...
+            '{store_name}'             => '',
         ];
 
+        add_action( 'dokan_reverse_withdrawal_invoice_email_sent', [ $this, 'trigger' ], 30, 2 );
+
         // Call parent constructor.
         parent::__construct();
 
-        $this->manual = true; // this email will be triggered manually
+        $this->manual = false; // this email will not be triggered manually, as it will be automatically triggered when async action processed.
```
Comment (nurul-umbhiya) [inline includes/Emails/ReverseWithdrawalInvoice.php:64]: you can safely remove this property

File: includes/Emails/VendorWithdrawRequest.php [hunk 2]
```diff
@@ -70,24 +80,16 @@ public function trigger( $user_id, $amount, $method ) {
...
-        $this->replace['amount']        = wc_price( $amount );
-        $this->replace['method']        = dokan_withdraw_get_method_title( $method );
-        $this->replace['profile_url']   = admin_url( 'user-edit.php?user_id=' . $seller->ID );
-        $this->replace['withdraw_page'] = admin_url( 'admin.php?page=dokan#/withdraw?status=pending' );
-        $this->replace['site_name']     = $this->get_from_name();
-        $this->replace['site_url']      = site_url();
-
         $this->setup_locale();
+        $this->object                          = $seller;
+        $this->placeholders['{user_name}']     = $seller->user_login;
```
Comment (nurul-umbhiya) [inline includes/Emails/VendorWithdrawRequest.php:85]: do we need to send user_login?

File: includes/Emails/VendorWithdrawRequest.php [hunk 3]
```diff
@@ -60,8 +69,9 @@ public function get_default_heading() {
...
      * Trigger the sending of this email.
      *
-     * @param int $product_id The product ID.
-     * @param array $postdata.
+     * @param int $user_id User ID.
+     * @param mixed $amount Withdrawal amount.
+     * @param string $method Withdrawal method.
      */
     public function trigger( $user_id, $amount, $method ) {
         $seller = get_user_by( 'id', $user_id );
```
Comment (nurul-umbhiya) [inline includes/Emails/VendorWithdrawRequest.php:77]: seller should be Vendor() instance

**Review by nurul-umbhiya — CHANGES_REQUESTED** [review]

**Review by nurul-umbhiya — CHANGES_REQUESTED** [review]

File: includes/Emails/ContactSeller.php [hunk 4]
```diff
@@ -27,12 +27,19 @@ class ContactSeller extends WC_Email {
...
+        $this->description    = __( 'These emails are sent to a vendor who is contacted by customer via contact form widget ', 'dokan-lite' );
+        $this->template_html  = 'emails/contact-seller.php';
+        $this->template_plain = 'emails/plain/contact-seller.php';
+        $this->template_base  = DOKAN_DIR . '/templates/';
+        $this->placeholders   = [
+            '{seller_name}'    => '',
+            '{customer_name}'  => '',
+            '{customer_email}' => '',
+            '{message}'        => '',
+            '{site_name}'      => $this->get_from_name(),
```
[the remark this answers is not in the document]
Comment (shohag121) [reply includes/Emails/ContactSeller.php:41]: {site_name} is needed.

File: includes/Emails/ContactSeller.php [hunk 5]
```diff
@@ -71,26 +78,16 @@ public function trigger( $seller_email, $contact_name, $contact_email, $contact_
         if ( ! $this->is_enabled() || ! $this->get_recipient() ) {
             return;
         }
+        $this->setup_locale();
 
         $this->from_email = $contact_email;
 
         $seller = get_user_by( 'email', $seller_email );
```
[the remark this answers is not in the document]
Comment (shohag121) [reply includes/Emails/ContactSeller.php:85]: Please elaborate bhai.

File: includes/Emails/NewProduct.php [hunk 6]
```diff
@@ -26,6 +27,15 @@ public function __construct() {
...
         $this->template_plain = 'emails/plain/new-product.php';
         $this->template_base  = DOKAN_DIR . '/templates/';
+        $this->placeholders = [
+            '{product_title}' => '',
+            '{price}'         => '',
+            '{seller_name}'   => '',
+            '{seller_url}'    => '',
+            '{category}'      => '',
+            '{product_link}'  => '',
+            '{site_name}'     => $this->get_from_name(),
```
[the remark this answers is not in the document]
Comment (shohag121) [reply includes/Emails/NewProduct.php:37]: {site_name} is needed.

File: includes/Emails/NewSeller.php [hunk 7]
```diff
@@ -18,12 +19,19 @@ class NewSeller extends WC_Email {
...
+        $this->description    = __( 'These emails are sent to chosen recipient(s) when a new vendor registers in marketplace', 'dokan-lite' );
+        $this->template_html  = 'emails/new-seller-registered.php';
+        $this->template_plain = 'emails/plain/new-seller-registered.php';
+        $this->template_base  = DOKAN_DIR . '/templates/';
+        $this->placeholders   = [
+            '{seller_name}' => '',
+            '{store_url}'   => '',
+            '{store_name}'  => '',
+            '{seller_edit}' => '',
+            '{site_name}'   => $this->get_from_name(),
```
[the remark this answers is not in the document]
Comment (shohag121) [reply includes/Emails/NewSeller.php:33]: {site_name} is needed.

File: includes/Emails/ProductPublished.php [hunk 8]
```diff
@@ -20,12 +21,20 @@ class ProductPublished extends WC_Email {
...
+        $this->template_html  = 'emails/product-published.php';
+        $this->template_plain = 'emails/plain/product-published.php';
+        $this->template_base  = DOKAN_DIR . '/templates/';
+        $this->placeholders   = [
+            '{product_title}'     => '',
+            '{price}'             => '',
+            '{seller_name}'       => '',
+            '{product_url}'       => '',
+            '{product_edit_link}' => '',
+            '{site_name}'         => $this->get_from_name(),
```
[the remark this answers is not in the document]
Comment (shohag121) [reply includes/Emails/ProductPublished.php:36]: {site_name} is needed.

File: templates/emails/plain/contact-seller.php [hunk 9]
```diff
@@ -17,11 +17,11 @@
 echo "\n=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=\n\n";
 
 // translators: 1) customer name 2) customer email
-echo sprintf( esc_html__( 'From : %1$s (%2$s)', 'dokan-lite' ), esc_html( $data['customer_name'] ), esc_html( $data['customer_email'] ) );
+echo sprintf( esc_html__( 'From : %1$s (%2$s)', 'dokan-lite' ), esc_html( $data['{customer_name}'] ), esc_html( $data['{customer_email}'] ) );
```
[the remark this answers is not in the document]
Comment (shohag121) [reply templates/emails/plain/contact-seller.php:20]: Yes, we do need `esc_html` here. 

https://github.com/woocommerce/woocommerce/blob/ceb5f942ea7f94c3df70635fbf8cd69a732588c8/plugins/woocommerce/templates/emails/plain/admin-cancelled-order.php#L25

**Review by nurul-umbhiya — APPROVED** [review]

**Comment by shohag121** [comment]
@shashwatahalder01 Bhai, Can you please test this PR?

**Comment by shashwatahalder01** [comment]
Scenario:
Placeholder `{message}` not showing for email type plain text.
Email Template: Dokan Contact Vendor.

Steps to reproduce:
1. customer fill dokan contact vendor from single store sidebar

Screenshot:
Plain Text:
Screenshot at Dec 21 16-48-52
Html:
Screenshot at Dec 22 16-31-29

**Comment by shashwatahalder01** [comment]
Scenario:
Message body is empty for email type plain text.
Email Template: Dokan New Product.

Steps to reproduce:
vendor creates a prouduct.

Screenshot:

Plain Text:
Screenshot at Dec 26 10-13-07

**Comment by shashwatahalder01** [comment]
Scenario:
No order details are showing for email type `plain text`.
Email Template: `Dokan Vendor New Order` 

Steps to reproduce:
customer place an order

Screenshot:

Plain Text:
Screenshot at Dec 26 12-06-45
HTML:
Screenshot at Dec 26 12-06-54

**Comment by shashwatahalder01** [comment]
Scenario:
Email description needs to be updated. Only the vendor is receiving the email but  `sent to chosen recipient` is mentioned in the description.
Email Template: `Dokan Vendor New Order`, `Dokan Vendor Completed Order`

Steps to reproduce:
1. Customer places an order.
2. Update order status to complete

Screenshot:
Screenshot at Dec 26 11-47-53
Screenshot at Dec 26 12-18-35

**Comment by shashwatahalder01** [comment]
Scenario:
No order details are showing for email type `plain text`.
Email Template: `Dokan Vendor Completed Order`

Steps to reproduce:
1.  update order status to complete
2. 
Screenshot:

Plain Text:
Screenshot at Dec 26 12-23-35

HTML:
Screenshot at Dec 26 12-23-49

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: `Dokan New Withdrawal Request`

- [ ]  1. profile link is missing on 
- [ ]  2. user_name is missing in two places
- [ ]  3. place holder should be `{store_name}` instead of {user_name}  cause value of {user_name} is store name not vendor user name.


Steps to reproduce:
request for withdraw as a vendor

Screenshot:

1
Plain Text:
Screenshot at Dec 26 12-55-03
HTML:
Screenshot at Dec 26 12-55-13

2.
Screenshot at Dec 26 12-43-09
Screenshot at Dec 26 12-43-35

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: `Dokan Withdraw Approved`
`{withdraw_page}` placeholder gives admin withdraw page URL, but the email is for vendor

 Steps to reproduce:
  1. Approve a withdrawal request
  
Screenshot:
Screenshot at Dec 26 14-56-41

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: `Dokan Withdraw Approved`
Content should be consistent for both `plain` and `HTML` email types.

Steps to reproduce:
1. Approve a withdrawal request

Screenshot:
Screenshot at Dec 26 14-51-48
Screenshot at Dec 26 15-59-56

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: `Dokan Vendor Product Review`
- [ ] 1. Description of Additional content should be in tooltip like all other email templates.
- [ ] 2. remove `and reply to` from email body as vendor can't reply to a review

Steps to reproduce:
review a product as a customer.

Screenshot:
1.
Screenshot at Dec 26 11-32-36

2.
Screenshot at Dec 26 11-32-36

**Comment by shashwatahalder01** [comment]
PHP Warnings:

- [ ] For email template: `Dokan Contact Vendor`

```
PHP Warning:  Undefined array key "customer_name" in /Users/rk/Sites/dokan3/wp-content/themes/storefront/dokan/emails/plain/contact-seller.php on line 20
PHP Warning:  Undefined array key "customer_email" in /Users/rk/Sites/dokan3/wp-content/themes/storefront/dokan/emails/plain/contact-seller.php on line 20
PHP Warning:  Undefined array key "message" in /Users/rk/Sites/dokan3/wp-content/themes/storefront/dokan/emails/plain/contact-seller.php on line 24
```

- [ ] For email template: `Dokan New Seller Registered`

```
PHP Warning:  Undefined array key "{site_name}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/new-seller-registered.php on line 20
```

- [ ]  For email template: `Dokan New Pending Product`


```
PHP Warning:  Undefined array key "site_url" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/new-product-pending.php on line 19
PHP Warning:  Undefined array key "site_name" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/new-product-pending.php on line 19
```

- [ ] For email template: `Dokan New Product`

```
PHP Warning:  Undefined array key "{site_name}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/new-product.php on line 19
```

- [ ] For email template: `Dokan New Withdrawal Request`

```
PHP Warning:  Undefined array key "{username}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/withdraw-new.php on line 20
PHP Warning:  Undefined array key "{username}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/withdraw-new.php on line 29
PHP Warning:  Undefined array key "{username}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/plain/withdraw-new.php on line 23
PHP Warning:  Undefined array key "{username}" in /Users/rk/Sites/dokan3/wp-content/plugins/dokan/templates/emails/plain/withdraw-new.php on line 35
```

**Comment by shashwatahalder01** [comment]
Scenario:

- [ ] {price} placeholder should follow woocommerce Number of decimals settings for email template:  `Dokan New Product`,

`Dokan New Pending Product `, `Dokan Pending Product Published`, 

- [ ] {amount} placeholder should follow woocommerce Number of decimals settings for email template:
`Dokan New Withdrawal Request`

Screenshot:
Screenshot at Dec 26 10-11-37

Screenshot at Dec 26 16-55-01

**Comment by shohag121** [comment]
> Scenario: Placeholder `{message}` not showing for email type plain text. Email Template: `Dokan Contact Vendor`.
> 
> Steps to reproduce:
> 
> 1. customer fills dokan contact vendor from single store sidebar
> 
> Screenshot: Plain Text: Screenshot at Dec 22 16-31-44
> 
> Html: Screenshot at Dec 22 16-31-29

Invalid issue.

**Comment by shohag121** [comment]
> Scenario: No order details are showing for email type `plain text`. Email Template: `Dokan Vendor New Order`
> 
> Steps to reproduce: customer place an order
> 
> Screenshot:
> 
> Plain Text: Screenshot at Dec 26 12-06-45 HTML: Screenshot at Dec 26 12-06-54

This template is not edited or changed by me in this PR. I don't know if this template is the reflection of any decision taken by product team.

**Comment by shohag121** [comment]
> Scenario: Email Template: `Dokan Withdraw Approved` Content should be consistent for both `plain` and `HTML` email types.
> 
> Steps to reproduce:
> 
> 1. Approve a withdrawal request
> 
> Screenshot: Screenshot at Dec 26 14-51-48 Screenshot at Dec 26 15-59-56
The text is changed after giving this PR. Don't know if the text change is intended or by mistake.

**Comment by shohag121** [comment]
> Dokan Vendor Product Review

This email class is added after this PR. It is not covered in this PR.

**Comment by shashwatahalder01** [comment]
Change `Username` to `StoreName`

Screenshot:
Screenshot at Jan 01 16-05-18

**Comment by shashwatahalder01** [comment]
Scenario:

- [ ] Price is not showing on both email type HTML and Plaintext.

- [ ] Both email types aren't consistent. In plain type there exist `click here` but link doesn't exists

 
Screenshot 2024-01-02 at 11 22 42 AM
Screenshot at Jan 02 11-18-42

**Comment by shashwatahalder01** [comment]
Scenario:
 Email template: `Dokan new product`
 

- [ ]  product edit link is missing in HTML format

- [ ]  correct the sentece

Screenshot at Jan 03 12-36-57

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: `Dokan New Withdrawal Request`
{User_name} placeholder is removed in this PR.  The user will see a `{User_name}` placeholder instead of a value after upgrading to the latest version.
We need to implement an upgrader For resolving {User_name} placeholder.



Screenshot:
Screenshot at Jan 03 12-46-29

**Comment by shashwatahalder01** [comment]
Scenario:
Multiple placeholder exists for the same value. 

 Site url -> {site_address}, {site_url}, 
 Store name -> {store_name} {seller_name}  {user_name}

**Comment by shashwatahalder01** [comment]
Scenario:
Email Template: Dokan Withdraw Approved
{profile_url} placeholder gives admin store edit page URL, but the email is for the vendor

Steps to reproduce:

Approve a withdrawal request
Screenshot:
Screenshot at Jan 04 12-41-13

**Pull request merged**